hugr-model/src/v0/mod.rs

```diff
@@ -0,0 +1,621 @@
+    #[inline]
+    pub fn get_node(&self, node_id: NodeId) -> Option<&Node<'a>> {
+        self.nodes.get(node_id.0 as usize)
+    }
+
+    /// Insert a new node into the module and return its id.
+    pub fn insert_node(&mut self, node: Node<'a>) -> NodeId {
+        let id = NodeId(self.nodes.len() as u32);
+        self.nodes.push(node);
+        id
+    }
+
+    /// Return the term data for a given term id.
+    #[inline]
+    pub fn get_term(&self, term_id: TermId) -> Option<&Term<'a>> {
+        self.terms.get(term_id.0 as usize)
+    }
+
+    /// Insert a new term into the module and return its id.
+    pub fn insert_term(&mut self, term: Term<'a>) -> TermId {
+        let id = TermId(self.terms.len() as u32);
+        self.terms.push(term);
+        id
+    }
+
+    /// Return the region data for a given region id.
+    #[inline]
+    pub fn get_region(&self, region_id: RegionId) -> Option<&Region<'a>> {
+        self.region.get(region_id.0 as usize)
+    }
+
+    /// Insert a new region into the module and return its id.
+    pub fn insert_region(&mut self, region: Region<'a>) -> RegionId {
+        let id = RegionId(self.region.len() as u32);
+        self.region.push(region);
+        id
+    }
+}
+
+/// Nodes in the hugr graph.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Node<'a> {
+    /// The operation that the node performs.
+    pub operation: Operation<'a>,
+    /// The input ports of the node.
+    pub inputs: &'a [Port<'a>],
+    /// The output ports of the node.
+    pub outputs: &'a [Port<'a>],
+    /// The parameters of the node.
+    pub params: &'a [TermId],
+    /// The regions of the node.
+    pub regions: &'a [RegionId],
+    /// The meta information attached to the node.
+    pub meta: &'a [MetaItem<'a>],
+    /// The type of the node.
+    pub r#type: TermId,
+}
+
+/// Operations that nodes can perform.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum Operation<'a> {
+    /// Data flow graphs.
+    Dfg,
+    /// Control flow graphs.
+    Cfg,
+    /// Basic blocks.
+    Block,
+    /// Function definitions.
+    DefineFunc {
+        /// The declaration of the function to be defined.
+        decl: &'a FuncDecl<'a>,
+    },
+    /// Function declarations.
+    DeclareFunc {
+        /// The function to be declared.
+        decl: &'a FuncDecl<'a>,
+    },
+    /// Function calls.
+    CallFunc {
+        /// The function to be called.
+        func: TermId,
+    },
+    /// Function constants.
+    LoadFunc {
+        /// The function to be loaded.
+        func: TermId,
+    },
+    /// Custom operation.
+    ///
+    /// The implicit parameters of the operation are left out.
+    Custom {
+        /// The name of the custom operation.
+        name: GlobalRef<'a>,
+    },
+    /// Custom operation.
+    ///
+    /// The implicit parameters of the operation are included.
+    CustomFull {
+        /// The name of the custom operation.
+        name: GlobalRef<'a>,
+    },
```

Comment:
why both?

Reply:
One only provides the explicit parameters, the other provides all parameters. Ultimately we want the one where all parameters are given, but that requires knowing the declaration of whatever thing the `GlobalRef` references, since we otherwise can't tell how many implicit parameters there are and where they are in the parameter list. A type inference step can convert the partial variant to the full variant since it has that information.

Also in the text format we can express both the full and the partial variant. Most of the time when you write these things manually you wouldn't want to include all implicit parameters, even if just as a wildcard. But rarely you might want to give them explicitly. This is inspired by how Lean does implicit parameters: https://docs.lean-lang.org/lean4/doc/implicit.html

The same happens in `Term::Apply` and `Term::ApplyFull`.

Instead of having these as different variants, we could have one variant with a flag. Matter of taste, I'd be happy with either.

Reply:
Thanks for the explanation - I'm happy with having different variants, would be nice to have a brief version of the above explanation in the docstrings